File: Lab 1/Task C/submission/bubble sort.cpp
#include <bits/stdc++.h>

using namespace std;

#define ll long long int
#define f(i,a,b) for(ll i=a;i<b;i++)

void bubbleSortNormal(ll *arr, ll n){
    f(i,0,n-1){
        int flag = 0;
        f(j,0,n-i-1){
            if(arr[j]>arr[j+1]){
                swap(arr[j],arr[j+1]);
                flag = 1;
            }
        }
        if(flag == 0) break;
    }
}

void bubbleSortReverse(ll *arr, ll n){
    f(i,0,n-1){
        int flag = 0;
        f(j,0,n-i-1){
            if(arr[j]<arr[j+1]){
                swap(arr[j],arr[j+1]);
                flag = 1;
            }
        }
        if(flag == 0) break;
    }
}

void bubbleSortAbs(ll *arr, ll n){
    f(i,0,n-1){
        int flag = 0;
        f(j,0,n-i-1){
            if((abs(arr[j])>abs(arr[j+1])) || ((abs(arr[j]) == abs(arr[j+1])) && (arr[j]>arr[j+1]))){
                swap(arr[j],arr[j+1]);
                flag = 1;
            }
        }
        if(flag == 0) break;
    }
}

int main(){
    ll n;
    cin>>n;
    ll arr[n];
    f(i,0,n) cin>>arr[i];
    bubbleSortNormal(arr,n);
    f(i,0,n) cout<<arr[i]<<" ";
    cout<<endl;
    bubbleSortReverse(arr, n);
    f(i,0,n) cout<<arr[i]<<" ";
    cout<<endl;
    bubbleSortAbs(arr,n);
    f(i,0,n) cout<<arr[i]<<" ";
    cout<<endl;
    return 0;
}

File: Lab 1/Task C/submission/bubble sort.py
def bubbleSortNorm (arr):
    n = len(arr)
    for i in range(n):
        flag = 0
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
                flag = 1
        if flag == 0:
            break

def bubbleSortRev (arr):
    n = len(arr)
    for i in range(n):
        flag = 0
        for j in range(0, n-i-1):
            if arr[j] < arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
                flag = 1
        if flag == 0:
            break

def bubbleSortAbs (arr):
    n = len(arr)
    for i in range(n):
        flag = 0
        for j in range(0, n-i-1):
            if abs(arr[j]) > abs(arr[j+1]) or (abs(arr[j]) == abs(arr[j+1]) and arr[j] > arr[j+1]):
                arr[j], arr[j+1] = arr[j+1], arr[j]
                flag = 1
        if flag == 0:
            break

n = int (input())
arr = list(map(int, input().split()))
bubbleSortNorm(arr)
print(*arr)
bubbleSortRev(arr)
print(*arr)
bubbleSortAbs(arr)
print(*arr)